### ISSUE TITLE
Incremental compilation bug with top-level definitions

### ISSUE THREAD

**Issue by mberndt123** [lampepfl/dotty#18447]
Hey,

I found a bug regarding incremental compilation in Scala 3. When you add a new Scala source file with a `def` that makes an existing function call ambiguous, this fact isn't detected by the incremental compiler and compilation erroneously succeeds when it should be failing. After a `clean`, compilation fails like it should.

To reproduce:
```
mkdir project
echo 'sbt.version=1.9.3' > project/build.properties
echo 'scalaVersion := "3.3.0"' > build.sbt

cat > A.scala << EOF
trait Foo
trait Bar

def foo(f: Foo) = ()
def qux = foo(new Foo with Bar {})
// this call is fine for now, but will be ambiguous once another overload is added
EOF
sbt compile
cat > B.scala << EOF
def foo(b: Bar) = () // add another overload, compilation should now fail
EOF
sbt compile # this succeeds, but shouldn't
sbt clean
sbt compile # now it fails, correctly
```

**Comment by som-snytt** [comment]
In the error message, I don't know what a "group" is.
```
Note that overloaded methods must all be defined in the same group of toplevel definitions
```
People love to point out the quirky English words for groups, such as, "a murder of crows". So this may be an opportunity to get creative.

**Comment by mberndt123** [comment]
Oh, tbh I hadn't ever really read the error message 😅 I just constructed the example so that it's ambiguous and assumed that that's what the error message is about. But in fact it doesn't seem to have anything to do with ambiguity. Oh well, it's still a bug that should be fixed.

Presumably "group" means that all overloads need to be defined in the same file…

**Comment by bishabosha** [comment]
is this another manifestation of https://github.com/sbt/zinc/issues/226?

**Comment by mberndt123** [comment]
Oh, tbh I hadn't ever really read the error message 😅 I just constructed the example so that it's ambiguous and assumed that that's what the error message is about. But in fact it seems to be illegal even if it didn't lead to ambiguity. Oh well, it's still a bug that needs to be fixed.

Presumably "group" just means that all overloads need to be defined in the same file then.

**Comment by mberndt123** [comment]
@bishabosha,

I doubt it, the zinc developers told me it's not their problem. 
https://github.com/sbt/zinc/issues/1239

**Comment by bishabosha** [comment]
Right, my point is that deciding to recompile A is a decision of Zinc, and relies upon A having a wildcard import of everything in the the empty package, which is known to be a bug

**Comment by bishabosha** [comment]
sure, sorry I didn't make it clear, this should happen for any package
